oeps/oep-0033-bp-code-autoformatting.rst

```diff
@@ -0,0 +1,127 @@
++-----------------+--------------------------------------------------------+
+| Title           | Code Autoformatting                                    |
++-----------------+--------------------------------------------------------+
+| Last Modified   | 2019-03-04                                             |
++-----------------+--------------------------------------------------------+
+| Authors         | Calen Pennington <cale@edx.org>                        |
++-----------------+--------------------------------------------------------+
+| Arbiter         | Ned Batchelder <ned@edx.org> (Tentatively)             |
++-----------------+--------------------------------------------------------+
+| Status          | Draft                                                  |
++-----------------+--------------------------------------------------------+
+| Type            | Best Practice                                          |
++-----------------+--------------------------------------------------------+
+| Created         | 2019-03-04                                             |
++-----------------+--------------------------------------------------------+
+| `Review Period` | 2019-03-04 - 2019-04-01                                |
++-----------------+--------------------------------------------------------+
+
+Context
+-------
+
+In a large software project or ecosystem like the Open edX codebase, consistency
+of code formatting makes it easier to read the code. In an ecosystem with many
+contributors, maintaining that consistency requires automated tools (such as
+linters), consistent code review, or both. In general, minimizing the number
+of review comments that a reviewer has to make on an incoming change makes
+review easier, and lessens the number of back-and-forths required to merge
+the code. As such, linters that can be run locally should be preferred.
+Similarly, ambiguity in style guidelines should be minimized, so that there
+are fewer places for commentary around what amounts to personal preference
+(which might then instigate unnecessary debate).
+
+In short, more consistent, less ambiguous standards for code formatting
+make for more productive code reviews and waste less time on inconsequential
+details.
+
+Decision
+--------
+
+We will select automatic code formatters for all of our major languages. These
+formatters will have both a lint-mode (that will be run in continuous
+integration to ensure consistent formatting) and an autoformat mode (which
+can be integrated with editors to free developers from having to think about
+code formatting). These formatting standards will be applied across all
+Open edX-authored repositories.
+
+Selected tools should, ideally, require minimal configuration. Any required
+configuration should be added to `edx/edx-lint`_ in order to allow it to
+be centralized and standardized across repositories in the Open edX ecosystem.
+
+The current recommended formatters are:
+
+    - Python: `black`_
```

Comment:
Last I looked into it, Black was too opinionated and barely, if at all configurable.
I found it too rigid. This would seem to maximize the amount of the codebase to be re-written,
which I'd like to avoid, per reasoning elsewhere in thread..

Reply:
So, opinionated is actually the goal, here. Whatever formatter is chosen should format any given code construct identically no matter how it's originally input. I think the pain of the rewrite is going to be there no matter how big or small it is (there's gonna be one big PR against the repo either way).

Reply:
> So, opinionated is actually the goal, here.

My hangup here isn't being opinionated;
it's being "poorly" opinionated.

I can rerun my test cases again if you'd like the specific violations,
but as mentioned, when I tried this on our code, `black` mandated changes that
I considered "wrong".

There where specific style patterns that we chose intentionally to make our fork more manageable, but we were no longer able to use these patterns under `black`
That's why I want flexibility/configuration; so we can forgo any overbearing rules.
Otherwise, we're getting hit on both the front and back ends;
- the upfront pain of the initial translation
- the ongoing cost of not being able to write code as we see fit for our fork

I'm not suggesting flexibility for the sake of endless and ongoing style debates
or a needlessly iterative process, but I don't want to be backed into a corner.
I want options if _needed_.

In RFC terms, think
`MAY reconfigure SOME options`
not
`WILL reconfigure MOST/ALL options`

> Whatever formatter is chosen should format any given code construct identically no matter how it's originally input.

Agreed.

> I think the pain of the rewrite is going to be there no matter how big or small it is

It's really a matter of perspective...
For edx, yeah, it's going to a pain so a single shot makes it easier.
Downstream, consider the following:
- _most_ side repos will be fairly painless for us;
  `platform` is my major concern re:transition
- So if you enable it for new directories, there's no pain downstream (or upstream, for that matter)

Now when it comes to running it against existing platform code, consider:
- single shot: for you
  - run the script
  - commit the code
  - move on
- single shot: for us
  - merge the code
  - spend weeks debugging/resolving conflicts because every code file in the platform has been modified 
  - hope we got it all
  - run the script against _our_ code now
  - commit the code
  - move on... to significant regression testing because _we_ didn't _just_ run an autoformating tool, we ran an autoformating tool, reworked virtually all of our custom integrations in every code file, then ran an autoformating tool again. It'll be more than a chore.

versus
- staged roll out: for you
  - run the script against a subset of additional directories
  - commit the code
  - repeat each named release until complete
- staged roll out: for us
  - merge the code
  - spend a much smaller amount of time resolving format-conflicts in the handful of newly-formatted subdirs
  - feel more confident that we got everything in the now smaller surface area
  - defer running the script against _our_ code now, so that we can deal w/ it out-of-band of this upgrade
  - spend significantly less time debugging and regression testing since the whole platform didn't change out from us over night, which allows us to get the named release live and not fall further behind, allowing us to 
  - repeat each named release until complete

I concede, single shot is easiest for you, it's one run of an automated tool.
Just consider how much more it ends up being for anyone with non-trivial platform modifications.
While a staged roll out would yield fewer commits, I'm not sure that's a selling point for us.
One commit `Auto format` isn't a practical difference for us versus multiple `Auto format 1/N`.

And a staged roll out let's us absorb those hits over time, ideally over the course of named releases. When there have been sweeping changes to the `edx-platform` codebase in the past,
having to "eat it" all at once basically drops our integration velocity down to zero for an extended period of time.

Reply:
> My hangup here isn't being opinionated;
> it's being "poorly" opinionated.

Yeah, that's fair enough. Like I said, I hadn't run into places where I objected w/ Black's formatting, so it would be helpful to me (while making a recommendation in this OEP) to see your specific counterexamples.

> Just consider how much more it ends up being for anyone with non-trivial platform modifications.
> While a staged roll out would yield fewer commits, I'm not sure that's a selling point for us.
> One commit `Auto format` isn't a practical difference for us versus multiple `Auto format 1/N`.
> 
> And a staged roll out let's us absorb those hits over time, ideally over the course of named releases. When there have been sweeping changes to the `edx-platform` codebase in the past,
> having to "eat it" all at once basically drops our integration velocity down to zero for an extended period of time.

I'm definitely willing to defer to your experience on this. It shouldn't be too much trouble to set up our integration tooling in a way that we can mark specific directories as having been `black`ened, and slowly expand that coverage. It's a bit cumbersome w/ editor auto-integration, but perhaps worth it. What pace of conversions is something that you think you and other integrators could manage? You mentioned being over the course of multiple open-edx releases, with puts the timeframe in months or years, which is a bit unfortunate from my POV.